File: elixir-phoenix/lib/rocket/engine/query.ex
defmodule Rocket.Engine.Query do
  @moduledoc "Builds parameterized SELECT/COUNT queries from query params."

  alias Rocket.Metadata.{Entity, Registry}

  defmodule QueryPlan do
    defstruct [:entity, filters: [], sorts: [], page: 1, per_page: 25, includes: []]
  end

  defmodule WhereClause do
    defstruct [:field, :operator, :value]
  end

  defmodule OrderClause do
    defstruct [:field, :dir]
  end

  defmodule QueryResult do
    defstruct [:sql, :params]
  end

  @doc "Parse Phoenix conn params into a QueryPlan."
  def parse_query_params(params, entity, registry) do
    plan = %QueryPlan{entity: entity, page: 1, per_page: 25}

    with {:ok, plan} <- parse_filters(plan, params, entity),
         {:ok, plan} <- parse_sorts(plan, params, entity),
         {:ok, plan} <- parse_pagination(plan, params),
         {:ok, plan} <- parse_includes(plan, params, entity, registry) do
      {:ok, plan}
    end
  end

  defp parse_filters(plan, params, entity) do
    filters =
      params
      |> Enum.filter(fn {key, _} -> String.starts_with?(key, "filter[") && String.ends_with?(key, "]") end)
      |> Enum.reduce_while([], fn {key, val}, acc ->
        inner = key |> String.slice(7..-2//1)
        {field, op} = parse_filter_key(inner)

        if Entity.has_field?(entity, field) do
          f = Entity.get_field(entity, field)

          case coerce_value(f, val, op) do
            {:ok, coerced} ->
              {:cont, [%WhereClause{field: field, operator: op, value: coerced} | acc]}

            {:error, msg} ->
              {:halt, {:error, Rocket.Engine.AppError.invalid_payload("Invalid filter value for #{field}: #{msg}")}}
          end
        else
          {:halt, {:error, Rocket.Engine.AppError.unknown_field("Unknown filter field: #{field}")}}
        end
      end)

    case filters do
      {:error, _} = err -> err
      list -> {:ok, %{plan | filters: Enum.reverse(list)}}
    end
  end

  defp parse_sorts(plan, params, entity) do
    case Map.get(params, "sort", "") do
      "" ->
        {:ok, plan}

      sort_param ->
        sorts =
          sort_param
          |> String.split(",")
          |> Enum.reduce_while([], fn part, acc ->
            part = String.trim(part)
            {dir, field} = if String.starts_with?(part, "-"), do: {"DESC", String.slice(part, 1..-1//1)}, else: {"ASC", part}

            if Entity.has_field?(entity, field) do
              {:cont, [%OrderClause{field: field, dir: dir} | acc]}
            else
              {:halt, {:error, Rocket.Engine.AppError.unknown_field("Unknown sort field: #{field}")}}
            end
          end)

        case sorts do
          {:error, _} = err -> err
          list -> {:ok, %{plan | sorts: Enum.reverse(list)}}
        end
    end
  end

  defp parse_pagination(plan, params) do
    page =
      case params["page"] do
        nil -> 1
        p -> max(parse_int(p, 1), 1)
      end

    per_page =
      case params["per_page"] do
        nil -> 25
        pp -> min(max(parse_int(pp, 25), 1), 100)
      end

    {:ok, %{plan | page: page, per_page: per_page}}
  end

  defp parse_includes(plan, params, entity, registry) do
    case Map.get(params, "include", "") do
      "" ->
        {:ok, plan}

      inc ->
        includes =
          inc
          |> String.split(",")
          |> Enum.reduce_while([], fn name, acc ->
            name = String.trim(name)
            rel = Registry.find_relation_for_entity(registry, name, entity.name)

            if rel do
              {:cont, [name | acc]}
            else
              {:halt, {:error, Rocket.Engine.AppError.unknown_field("Unknown include: #{name}")}}
            end
          end)

        case includes do
          {:error, _} = err -> err
          list -> {:ok, %{plan | includes: Enum.reverse(list)}}
        end
    end
  end

  @doc "Build parameterized SELECT SQL."
  def build_select_sql(%QueryPlan{} = plan) do
    {params, n} = {[], 0}
    entity = plan.entity

    columns = entity |> Entity.field_names() |> Enum.join(", ")

    columns =
      if entity.soft_delete && !Entity.has_field?(entity, "deleted_at") do
        columns <> ", deleted_at"
      else
        columns
      end

    {where, params, n} = build_where_clauses(plan, params, n)

    sql = "SELECT #{columns} FROM #{entity.table}"
    sql = if where != "", do: sql <> " WHERE " <> where, else: sql

    # Sort
    sql =
      if plan.sorts != [] do
        order_parts = Enum.map(plan.sorts, fn s -> "#{s.field} #{s.dir}" end)
        sql <> " ORDER BY " <> Enum.join(order_parts, ", ")
      else
        sql
      end

    # Pagination
    n = n + 1
    params = params ++ [plan.per_page]
    limit_ph = "$#{n}"
    n = n + 1
    params = params ++ [(plan.page - 1) * plan.per_page]
    offset_ph = "$#{n}"
    sql = sql <> " LIMIT #{limit_ph} OFFSET #{offset_ph}"

    %QueryResult{sql: sql, params: params}
  end

  @doc "Build parameterized COUNT SQL."
  def build_count_sql(%QueryPlan{} = plan) do
    {where, params, _n} = build_where_clauses(plan, [], 0)

    sql = "SELECT COUNT(*) as count FROM #{plan.entity.table}"
    sql = if where != "", do: sql <> " WHERE " <> where, else: sql

    %QueryResult{sql: sql, params: params}
  end

  defp build_where_clauses(plan, params, n) do
    entity = plan.entity

    {clauses, params, n} =
      if entity.soft_delete do
        {["deleted_at IS NULL"], params, n}
      else
        {[], params, n}
      end

    {clauses, params, n} =
      Enum.reduce(plan.filters, {clauses, params, n}, fn f, {cls, prms, idx} ->
        {clause, prms, idx} = build_where_clause(f, prms, idx)
        {cls ++ [clause], prms, idx}
      end)

    where = Enum.join(clauses, " AND ")
    {where, params, n}
  end

  defp build_where_clause(%WhereClause{operator: "in"} = f, params, n) do
    dialect = Rocket.Store.dialect()
    {clause, extra_params, n} = dialect.in_expr(f.field, f.value, n)
    {clause, params ++ extra_params, n}
  end

  defp build_where_clause(%WhereClause{operator: "not_in"} = f, params, n) do
    dialect = Rocket.Store.dialect()
    {clause, extra_params, n} = dialect.not_in_expr(f.field, f.value, n)
    {clause, params ++ extra_params, n}
  end

  defp build_where_clause(%WhereClause{} = f, params, n) do
    n = n + 1
    ph = "$#{n}"
    params = params ++ [f.value]

    clause =
      case f.operator do
        "eq" -> "#{f.field} = #{ph}"
        "neq" -> "#{f.field} != #{ph}"
        "gt" -> "#{f.field} > #{ph}"
        "gte" -> "#{f.field} >= #{ph}"
        "lt" -> "#{f.field} < #{ph}"
        "lte" -> "#{f.field} <= #{ph}"
        "like" -> "#{f.field} LIKE #{ph}"
        _ -> "#{f.field} = #{ph}"
      end

    {clause, params, n}
  end

  defp parse_filter_key(key) do
    case String.split(key, ".", parts: 2) do
      [field, op] -> {field, op}
      [field] -> {field, "eq"}
    end
  end

  defp coerce_value(field, val, op) when op in ["in", "not_in"] do
    parts = String.split(val, ",")

    results =
      Enum.map(parts, fn p ->
        coerce_single_value(field, String.trim(p))
      end)

    case Enum.find(results, &match?({:error, _}, &1)) do
      nil -> {:ok, Enum.map(results, fn {:ok, v} -> v end)}
      err -> err
    end
  end

  defp coerce_value(field, val, _op), do: coerce_single_value(field, val)

  defp coerce_single_value(%{type: "int"}, val) do
    case Integer.parse(val) do
      {n, ""} -> {:ok, n}
      _ -> {:error, "expected integer"}
    end
  end

  defp coerce_single_value(%{type: "bigint"}, val) do
    case Integer.parse(val) do
      {n, ""} -> {:ok, n}
      _ -> {:error, "expected integer"}
    end
  end

  defp coerce_single_value(%{type: "decimal"}, val) do
    case Float.parse(val) do
      {n, ""} -> {:ok, n}
      _ -> {:error, "expected number"}
    end
  end

  defp coerce_single_value(%{type: "boolean"}, val) do
    case val do
      v when v in ["true", "1"] -> {:ok, true}
      v when v in ["false", "0"] -> {:ok, false}
      _ -> {:error, "expected boolean"}
    end
  end

  defp coerce_single_value(_, val), do: {:ok, val}

  defp parse_int(val, default) when is_binary(val) do
    case Integer.parse(val) do
      {n, ""} -> n
      _ -> default
    end
  end

  defp parse_int(val, _) when is_integer(val), do: val
  defp parse_int(_, default), do: default
end

File: backend/golang/internal/engine/query.go
package engine

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/gofiber/fiber/v2"

	"rocket-backend/internal/metadata"
	"rocket-backend/internal/store"
)

type QueryPlan struct {
	Entity   *metadata.Entity
	Filters  []WhereClause
	Sorts    []OrderClause
	Page     int
	PerPage  int
	Includes []string
}

type WhereClause struct {
	Field    string
	Operator string
	Value    any
}

type OrderClause struct {
	Field string
	Dir   string // ASC or DESC
}

type QueryResult struct {
	SQL    string
	Params []any
}

// ParseQueryParams parses Fiber query parameters into a QueryPlan.
func ParseQueryParams(c *fiber.Ctx, entity *metadata.Entity, reg *metadata.Registry) (*QueryPlan, error) {
	plan := &QueryPlan{
		Entity:  entity,
		Page:    1,
		PerPage: 25,
	}

	// Parse filters: filter[field]=val or filter[field.op]=val
	queries := c.Queries()
	for key, val := range queries {
		if !strings.HasPrefix(key, "filter[") || !strings.HasSuffix(key, "]") {
			continue
		}
		inner := key[7 : len(key)-1] // extract between [ and ]
		field, op := parseFilterKey(inner)

		if !entity.HasField(field) {
			return nil, &AppError{
				Code:    "UNKNOWN_FIELD",
				Status:  400,
				Message: fmt.Sprintf("Unknown filter field: %s", field),
			}
		}

		coerced, err := coerceValue(entity.GetField(field), val, op)
		if err != nil {
			return nil, &AppError{
				Code:    "INVALID_PAYLOAD",
				Status:  400,
				Message: fmt.Sprintf("Invalid filter value for %s: %v", field, err),
			}
		}

		plan.Filters = append(plan.Filters, WhereClause{
			Field:    field,
			Operator: op,
			Value:    coerced,
		})
	}

	// Parse sort: sort=-created_at,name
	if sortParam := c.Query("sort"); sortParam != "" {
		parts := strings.Split(sortParam, ",")
		for _, part := range parts {
			part = strings.TrimSpace(part)
			dir := "ASC"
			field := part
			if strings.HasPrefix(part, "-") {
				dir = "DESC"
				field = part[1:]
			}
			if !entity.HasField(field) {
				return nil, &AppError{
					Code:    "UNKNOWN_FIELD",
					Status:  400,
					Message: fmt.Sprintf("Unknown sort field: %s", field),
				}
			}
			plan.Sorts = append(plan.Sorts, OrderClause{Field: field, Dir: dir})
		}
	}

	// Parse pagination
	if p := c.Query("page"); p != "" {
		if v, err := strconv.Atoi(p); err == nil && v > 0 {
			plan.Page = v
		}
	}
	if pp := c.Query("per_page"); pp != "" {
		if v, err := strconv.Atoi(pp); err == nil && v > 0 {
			plan.PerPage = v
			if plan.PerPage > 100 {
				plan.PerPage = 100
			}
		}
	}

	// Parse includes: include=items,customer
	if inc := c.Query("include"); inc != "" {
		parts := strings.Split(inc, ",")
		for _, name := range parts {
			name = strings.TrimSpace(name)
			rel := reg.FindRelationForEntity(name, entity.Name)
			if rel == nil {
				return nil, &AppError{
					Code:    "UNKNOWN_FIELD",
					Status:  400,
					Message: fmt.Sprintf("Unknown include: %s", name),
				}
			}
			plan.Includes = append(plan.Includes, name)
		}
	}

	return plan, nil
}

// BuildSelectSQL builds a parameterized SELECT statement from the query plan.
func BuildSelectSQL(plan *QueryPlan, dialect store.Dialect) QueryResult {
	pb := dialect.NewParamBuilder()
	entity := plan.Entity

	columns := strings.Join(entity.FieldNames(), ", ")
	if entity.SoftDelete && entity.GetField("deleted_at") == nil {
		columns += ", deleted_at"
	}

	var where []string

	// Soft delete filter
	if entity.SoftDelete {
		where = append(where, "deleted_at IS NULL")
	}

	// User filters
	for _, f := range plan.Filters {
		clause := buildWhereClause(f, pb, dialect)
		where = append(where, clause)
	}

	sql := fmt.Sprintf("SELECT %s FROM %s", columns, entity.Table)
	if len(where) > 0 {
		sql += " WHERE " + strings.Join(where, " AND ")
	}

	// Sort
	if len(plan.Sorts) > 0 {
		var orderParts []string
		for _, s := range plan.Sorts {
			orderParts = append(orderParts, fmt.Sprintf("%s %s", s.Field, s.Dir))
		}
		sql += " ORDER BY " + strings.Join(orderParts, ", ")
	}

	// Pagination
	limit := pb.Add(plan.PerPage)
	offset := pb.Add((plan.Page - 1) * plan.PerPage)
	sql += fmt.Sprintf(" LIMIT %s OFFSET %s", limit, offset)

	return QueryResult{SQL: sql, Params: pb.Params()}
}

// BuildCountSQL builds a COUNT query with the same filters as the select.
func BuildCountSQL(plan *QueryPlan, dialect store.Dialect) QueryResult {
	pb := dialect.NewParamBuilder()
	entity := plan.Entity

	var where []string
	if entity.SoftDelete {
		where = append(where, "deleted_at IS NULL")
	}
	for _, f := range plan.Filters {
		clause := buildWhereClause(f, pb, dialect)
		where = append(where, clause)
	}

	sql := fmt.Sprintf("SELECT COUNT(*) FROM %s", entity.Table)
	if len(where) > 0 {
		sql += " WHERE " + strings.Join(where, " AND ")
	}

	return QueryResult{SQL: sql, Params: pb.Params()}
}

func buildWhereClause(f WhereClause, pb store.ParamBuilder, dialect store.Dialect) string {
	switch f.Operator {
	case "eq", "":
		return fmt.Sprintf("%s = %s", f.Field, pb.Add(f.Value))
	case "neq":
		return fmt.Sprintf("%s != %s", f.Field, pb.Add(f.Value))
	case "gt":
		return fmt.Sprintf("%s > %s", f.Field, pb.Add(f.Value))
	case "gte":
		return fmt.Sprintf("%s >= %s", f.Field, pb.Add(f.Value))
	case "lt":
		return fmt.Sprintf("%s < %s", f.Field, pb.Add(f.Value))
	case "lte":
		return fmt.Sprintf("%s <= %s", f.Field, pb.Add(f.Value))
	case "in":
		values, ok := f.Value.([]any)
		if !ok {
			return fmt.Sprintf("%s = %s", f.Field, pb.Add(f.Value))
		}
		return dialect.InExpr(f.Field, pb, values)
	case "not_in":
		values, ok := f.Value.([]any)
		if !ok {
			return fmt.Sprintf("%s != %s", f.Field, pb.Add(f.Value))
		}
		return dialect.NotInExpr(f.Field, pb, values)
	case "like":
		return fmt.Sprintf("%s LIKE %s", f.Field, pb.Add(f.Value))
	default:
		return fmt.Sprintf("%s = %s", f.Field, pb.Add(f.Value))
	}
}

// parseFilterKey splits "total.gte" into ("total", "gte") or "status" into ("status", "eq").
func parseFilterKey(key string) (string, string) {
	parts := strings.SplitN(key, ".", 2)
	if len(parts) == 2 {
		return parts[0], parts[1]
	}
	return key, "eq"
}

// coerceValue converts string query param values to appropriate Go types based on field metadata.
func coerceValue(field *metadata.Field, val string, op string) (any, error) {
	// Handle "in" and "not_in" as comma-separated arrays
	if op == "in" || op == "not_in" {
		parts := strings.Split(val, ",")
		coerced := make([]any, len(parts))
		for i, p := range parts {
			v, err := coerceSingleValue(field, strings.TrimSpace(p))
			if err != nil {
				return nil, err
			}
			coerced[i] = v
		}
		return coerced, nil
	}

	return coerceSingleValue(field, val)
}

func coerceSingleValue(field *metadata.Field, val string) (any, error) {
	switch field.Type {
	case "int":
		return strconv.Atoi(val)
	case "bigint":
		return strconv.ParseInt(val, 10, 64)
	case "decimal":
		return strconv.ParseFloat(val, 64)
	case "boolean":
		return strconv.ParseBool(val)
	default:
		return val, nil
	}
}
